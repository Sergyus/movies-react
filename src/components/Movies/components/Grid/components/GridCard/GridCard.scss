@import "../../../../../../styles/variables";

.card {
  height: 380px;
  color: white;
  background: white;
  border-radius: 5px;
  position: relative;
  box-shadow: 0 10px 40px 0 rgba(0, 0, 0, 0.1);

  & .onHover {
    display: none;
    position: absolute;
    top: 0;
    right: 0;
    width: 100%;
    height: 70%;
    background-color: rgba(0, 0, 0, 0.7);
  }

  &:hover .onHover {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
  }

  & img {
    object-fit: cover;
    border-top-left-radius: 5px;
    border-top-right-radius: 5px;
    width: 100%;
    transition: height 0.3s;
  }
}

.btnWatch {
  cursor: pointer;
  height: 100px;
  width: 100px;
  outline: none;
  background-color: transparent;
  color: white;
  border: none;
  margin-bottom: 50px;
  transition: transform 0.3s;
  background-image: url("../../../../../../assets/images/sprites.png");
  background-position: -20px -30px;
  font-size: 1rem;

  & span {
    display: inline-block;
    margin-top: 90px;
  }

  &:hover {
    transform: scale(1.05);
  }
}

.info {
  padding: 0 15px;
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  align-items: start;
  position: relative;
  margin-top: 13px;
  color: black;

  & > div {
    text-align: left;
    width: 75%;

    & > p {
      display: flex;
      flex-wrap: wrap;
      font-size: 0.8rem;
      line-height: 1.2rem;
      color: $blue-color;
      font-family: "Helvetica Neue", Arial, Helvetica, sans-serif;
      font-weight: bold;
    }
  }

  & > span {
    color: $blue-color;
    font-size: 0.9rem;
  }
}

.title {
  color: #242b2d;
  margin-bottom: 5px;
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

.overlay {
  background-color: rgba(0, 0, 0, 0.8);
  position: absolute;
  top: 0;
  width: 100%;
  height: 100%;
  text-align: left;

  .title {
    color: white;
    margin-bottom: 5px;
    text-overflow: ellipsis;
    overflow: hidden;
    white-space: nowrap;
  }
}

.close {
  cursor: pointer;
  background-color: transparent;
  border: none;
  outline: none;
  position: absolute;
  top: 20px;
  left: 15px;
  content: "";
  height: 20px;
  width: 20px;
  background-image: url("../../../../../../assets/images/sprites.png");
  background-position: -250px -30px;
  transition: transform 0.3s;

  &:hover {
    transform: scale(1.2);
  }
}

.wrapper {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  align-items: start;
  margin-top: 60px;
  padding: 5px 15px;
  position: relative;
  margin-bottom: 15px;
  color: $blue-color;
  font-family: "Helvetica Neue", Arial, Helvetica, sans-serif;
  font-weight: bold;

  & > span {
    font-size: 0.9rem;
  }

  & > div {
    max-width: 80%;
    display: flex;
    flex-wrap: wrap;

    & > .title {
      width: 100%;

      & ~ span {
        font-size: 0.8rem;
        line-height: 1.2rem;
      }
    }
  }
}

.description {
  padding: 0 20px;
  font-size: 0.9rem;
  margin-bottom: 30px;
  line-height: 1.5rem;

  & + button {
    position: absolute;
    bottom: 20px;
    margin-left: 20px;
  }
}

@media (max-width: 900px) {
  .videoContainer {
    padding: 2.5em 0;
  }

  .video {
    width: 90%;
    height: 60%;
    margin: 0 auto;
  }

  .closeVideo {
    top: 20%;
    right: 2%;
  }
}
