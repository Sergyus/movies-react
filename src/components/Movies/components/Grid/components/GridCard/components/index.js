import OverlayCard from './OverlayCard';
import DefaultCard from './DefaultCard';

export { OverlayCard, DefaultCard };
