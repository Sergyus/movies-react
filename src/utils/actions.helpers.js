export const createAsyncFnRequest = requestName => ({
  type: requestName,
});

export const createAsyncFnFailure = (requestName, error) => ({
  type: requestName,
  error: error.message,
});
