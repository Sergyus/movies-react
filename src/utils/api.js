const MOVIE_DB_API_KEY = "ed98d25919802b5020e8feb6c3923886";
const MOVIE_DB_BASE_URL = "https://api.themoviedb.org/3";
const LANGUAGE = "ru-RU";

const createMovieDbUrl = (relativeUrl, queryParams) => {
  let baseUrl = `${MOVIE_DB_BASE_URL}${relativeUrl}?api_key=${MOVIE_DB_API_KEY}&language=${LANGUAGE}`;
  if (queryParams) {
    Object.keys(queryParams).forEach(paramName => {
      baseUrl += `&${paramName}=${queryParams[paramName]}`;
      return baseUrl;
    });
  }
  return baseUrl;
};

const getLatestMovies = ({ page }) => {
  const fullUrl = createMovieDbUrl("/movie/popular", {
    page
  });
  return fetch(fullUrl);
};

const getTopMovies = ({ page }) => {
  const fullUrl = createMovieDbUrl("/movie/top_rated", {
    page
  });
  return fetch(fullUrl);
};

const getUpcomingMovies = ({ page }) => {
  const fullUrl = createMovieDbUrl("/movie/upcoming", {
    page
  });
  return fetch(fullUrl);
};

const searchMovies = ({ page, query }) => {
  const fullUrl = `${createMovieDbUrl("/search/movie", {
    page,
    query
  })}&include_adult=false`;
  return fetch(fullUrl);
};

const getGenres = () => {
  const fullUrl = createMovieDbUrl("/genre/movie/list");
  return fetch(fullUrl);
};

const getMoviesByGenre = ({ page, genre }) => {
  const fullUrl = `${createMovieDbUrl("/discover/movie", {
    page
  })}&sort_by=popularity.desc&include_adult=false&include_video=false&with_genres=${genre}`;
  return fetch(fullUrl);
};

const getMovieDetails = ({ movieId }) => {
  const fullUrl = createMovieDbUrl(`/movie/${movieId}`);
  return fetch(fullUrl);
};

const getMovieVideo = movieId => {
  const fullUrl = createMovieDbUrl(`/movie/${movieId}/videos`);
  return fetch(fullUrl);
};

export {
  getTopMovies,
  searchMovies,
  getMovieDetails,
  getGenres,
  getMoviesByGenre,
  getLatestMovies,
  getUpcomingMovies,
  getMovieVideo
};
