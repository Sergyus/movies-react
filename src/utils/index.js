import {
  getTopMovies, searchMovies, getGenres, getMoviesByGenre,
  getLatestMovies, getUpcomingMovies, getMovieVideo,
} from './api';
import { getMovies } from './tbdbApiService';
import truncate from './truncate';
import { useFetchOnScroll } from './useFetchOnScroll';
import useBodyScrollLock from './useBodyScrollLock';
import useOnClickOutside from './useOnClickOutside';
import { createAsyncFnFailure, createAsyncFnRequest } from './actions.helpers';

export {
  getTopMovies, searchMovies, getGenres, getMoviesByGenre,
  getLatestMovies, getUpcomingMovies, getMovieVideo,
  truncate, useBodyScrollLock, useOnClickOutside, getMovies,
  createAsyncFnFailure, createAsyncFnRequest, useFetchOnScroll,
};
