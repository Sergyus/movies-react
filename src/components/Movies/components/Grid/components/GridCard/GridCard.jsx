import React from 'react';
import PropTypes from 'prop-types';
import styles from './GridCard.scss';
import { OverlayCard, DefaultCard } from './components';

const GridCard = ({
  movie, isActive, changeActiveCard,
  toggleModal, fetchVideoSrc,
}) => {
  const {
    rating,
    title,
    genres,
    overview,
    image,
    id,
  } = movie;
  return (
    <article className={styles.card}>
      <img height={isActive ? '70%' : '100%'} src={image} alt="Poster" />
      { isActive
        ? (
          <DefaultCard
            id={id}
            title={title}
            rating={rating}
            changeActiveCard={changeActiveCard}
            genres={genres}
            toggleModal={toggleModal}
            fetchVideoSrc={fetchVideoSrc}
          />
        )
        : (
          <OverlayCard
            id={id}
            overview={overview}
            title={title}
            rating={rating}
            changeActiveCard={changeActiveCard}
            genres={genres}
            fetchVideoSrc={fetchVideoSrc}
            toggleModal={toggleModal}
          />
        )}
    </article>
  );
};

GridCard.propTypes = {
  movie: PropTypes.shape({
    rating: PropTypes.number,
    title: PropTypes.string,
    genres: PropTypes.array,
    overview: PropTypes.string,
    image: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  }).isRequired,
  isActive: PropTypes.bool.isRequired,
  changeActiveCard: PropTypes.func.isRequired,
  toggleModal: PropTypes.func,
  fetchVideoSrc: PropTypes.func,
};

GridCard.defaultProps = {
  toggleModal: () => {},
  fetchVideoSrc: () => {},
};

export default GridCard;
