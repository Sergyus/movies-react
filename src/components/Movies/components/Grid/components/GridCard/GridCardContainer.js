import { connect } from 'react-redux';
import GridCard from './GridCard';
import { fetchVideoSrc } from '../../../../../../modules/app';

export default connect(null, { fetchVideoSrc })(GridCard);
