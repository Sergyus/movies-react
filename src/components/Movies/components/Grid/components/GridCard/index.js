import GridCardContainer from './GridCardContainer';

export default GridCardContainer;
