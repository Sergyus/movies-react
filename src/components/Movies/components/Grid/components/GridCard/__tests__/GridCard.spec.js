import React, { useState } from 'react';
import { create, act } from 'react-test-renderer';
import GridCard from '../GridCard';
import { getMovies } from '../../../../../../../utils';
import mockMovies from '../../../../../../../assets/json/popular.json';
import { genres } from '../../../../../../../assets/json/genres.json';

describe('GridCard component renders correctly', () => {
  const movie = { ...getMovies(mockMovies.results, genres)[0], id: 1 };
  const activeCard = 1;
  const mockFunction = jest.fn();

  it('GridCard renders and works correctly', () => {
    const tree = create(
      <GridCard
        movie={movie}
        activeCard={activeCard}
        changeActiveCard={mockFunction}
        isActive={false}
      />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Set overlay works correctly', () => {
    const Wrapper = () => {
      const film = { ...getMovies(mockMovies.results, genres)[0], id: 2 };
      const [currentCard, setCurrentCard] = useState(1);
      return (
        <>
          <GridCard
            movie={film}
            changeActiveCard={id => setCurrentCard(id)}
            isActive={film.id !== currentCard}
          />
        </>
      );
    };
    const tree = create(
      <Wrapper />,
    );
    const watchBtn = tree.root.findByProps({ 'aria-label': 'Watch' });
    const mockWatchBtn = jest.fn(watchBtn.props.onClick);
    const button = tree.root.findByProps({ 'aria-label': 'View Info' });
    act(() => {
      mockWatchBtn();
      button.props.onClick();
    });
    const overlay = tree.toJSON().children[1];
    expect(mockWatchBtn).toHaveBeenCalled();
    expect(overlay.props).toEqual({ className: 'overlay' });
  });

  it('Set default card with short overview works correctly', () => {
    const Wrapper = () => {
      const film = { ...getMovies(mockMovies.results, genres)[0], id: 2 };
      const [currentCard, setCurrentCard] = useState(2);
      return (
        <>
          <GridCard
            movie={film}
            changeActiveCard={id => setCurrentCard(id)}
            isActive={film.id !== currentCard}
          />
        </>
      );
    };
    const tree = create(
      <Wrapper />,
    );
    const watchBtn = tree.root.findByProps({ 'aria-label': 'Watch Now' });
    const mockWatchBtn = jest.fn(watchBtn.props.onClick);
    const button = tree.root.findByProps({ 'aria-label': 'Close' });
    act(() => {
      mockWatchBtn();
      button.props.onClick();
    });
    const onHover = tree.toJSON().children[1];
    expect(onHover.props).toEqual({ className: 'onHover' });
    expect(mockWatchBtn).toHaveBeenCalled();
  });
});
