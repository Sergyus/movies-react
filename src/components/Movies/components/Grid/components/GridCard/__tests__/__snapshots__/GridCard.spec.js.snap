// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridCard component renders correctly GridCard renders and works correctly 1`] = `
<article
  className="card"
>
  <img
    alt="Poster"
    height="100%"
    src="https://image.tmdb.org/t/p/w300/AnLCqcH66Ecu8VFmXu3RofK4SMK.jpg"
  />
  <div
    className="overlay"
  >
    <button
      aria-label="Close"
      className="close"
      onClick={[Function]}
      type="button"
    />
    <div
      className="wrapper"
    >
      <div>
        <h3
          className="title"
        >
          T-34
        </h3>
        <span
          className="genre"
        >
          War
        </span>
        <span
          className="genre"
        >
          Drama
        </span>
        <span
          className="genre"
        >
          Adventure
        </span>
        <span
          className="genre"
        >
          Action
        </span>
      </div>
      <span
        className="numRating"
      >
        4.5
      </span>
    </div>
    <p
      className="description"
    >
      1941 - WWii. the second lieutenant Nikolai ivushkin, commander of a t-34, engages in an unequal battle against the tank ace Klaus Jager in a battle near moscow. His mission is more...
    </p>
    <button
      aria-label="Watch Now"
      className="btn primary"
      onClick={[Function]}
      type="button"
    >
      Watch Now
    </button>
  </div>
</article>
`;
